src/framework/results/experiment_data.hpp

```diff
@@ -298,84 +379,87 @@ ExperimentData& ExperimentData::combine(ExperimentData &data) {
-    singleshot_snapshots_[pair.first].combine(pair.second);
+  for (auto &pair : data.pershot_json_snapshots_) {
+    pershot_json_snapshots_[pair.first].combine(pair.second);
   }
-  for (auto &pair : data.average_snapshots_) {
-    average_snapshots_[pair.first].combine(pair.second);
+  for (auto &pair : data.average_json_snapshots_) {
+    average_json_snapshots_[pair.first].combine(pair.second);
   }
   // Combine metadata
-  for(auto& metadatum: data.metadata_.items()) {
-    metadata_[metadatum.key()] = metadatum.value();
+  for (auto &pair : data.metadata_) {
+    metadata_[pair.first] = pair.second;
   }
+
   // Combine additional data
-  // Note that this will override any fields that have the same value
-  for (auto it = data.additional_data_.begin();
-       it != data.additional_data_.end(); ++it) {
-    // Check if key exists
-    if (additional_data_.find(it.key()) == additional_data_.end()) {
-      // If it doesn't add the data
-      additional_data_[it.key()] = it.value();
-    } else {
-      // If it does we append the data.
-      // Note that this will overwrite any subkeys with same name
-      for (auto it2 = it.value().begin(); it2 != it.value().end(); it2++) {
-        additional_data_[it.key()][it2.key()] = it2.value();
-      }
-    }
+  for (auto &pair : data.additional_json_data_) {
+    const auto &key = pair.first;
+    erase_additional_data(key);
+    additional_json_data_[key] = std::move(pair.second);
+  }
+  for (auto &pair : data.additional_cvector_data_) {
+    const auto &key = pair.first;
+    erase_additional_data(key);
+    additional_cvector_data_[key] = std::move(pair.second);
+  }
+  for (auto &pair : data.additional_cmatrix_data_) {
+    const auto &key = pair.first;
+    erase_additional_data(key);
+    additional_cmatrix_data_[key] = std::move(pair.second);
   }
-
   // Clear any remaining data from other container
   data.clear();
 
   return *this;
 }
 
+//------------------------------------------------------------------------------
+// JSON serialization
+//------------------------------------------------------------------------------
 
 json_t ExperimentData::json() const {
-
   // Initialize output as additional data JSON
-  json_t tmp = additional_data_;
-
-  // Add standard data
-  // This will override any additional data fields if they use keys:
-  // "counts", "memory", "register", "snapshots"
+  json_t tmp;
 
   // Measure data
-  if (return_counts_ && counts_.empty() == false)
-    tmp["counts"] = counts_;
-  if (return_memory_ && memory_.empty() == false)
-    tmp["memory"] = memory_;
+  if (return_counts_ && counts_.empty() == false) tmp["counts"] = counts_;
+  if (return_memory_ && memory_.empty() == false) tmp["memory"] = memory_;
   if (return_register_ && register_.empty() == false)
     tmp["register"] = register_;
-  // Average snapshot data
+
+  // Add additional data
+  for (const auto &pair : additional_json_data_) {
+    tmp[pair.first] = pair.second;
```

Comment:
If `json_t` type implements move constructor, I guess we can use move semantics here to avoid copying:
```suggestion
    tmp[pair.first] = std::move(pair.second);
```

Reply:
We might want to double check that `json_t` (a.k.a: `nlohmann::json`) actually has a move constructor/assignament operator.

Reply:
The serialization method has to be `const` on original object so can't use move anyway